### DISCUSSION TITLE
Template strings have some parsing bugs

### DISCUSSION THREAD

**Discussion by I-Want-ToBelieve** [google/zx#700] [Help needed] [answered]
"zx": "^7.2.3"

```ts
// Define the signals
const signals = [
  { event: 'application_launched', label: 'yakite-application-launched', action: 'event application-launched --env YABAI_PROCESS_ID=$YABAI_PROCESS_ID' },
  { event: 'application_terminated', label: 'yakite-application-terminated', action: 'event application-terminated --env YABAI_PROCESS_ID=$YABAI_PROCESS_ID' }
]

  for (const { event, label, action } of signals) {
    console.log(`action=${action}`)  //  action=event application-launched --env YABAI_PROCESS_ID=$YABAI_PROCESS_ID
    $`action=${action}` // action=$'event application-launched --env YABAI_PROCESS_ID=$YABAI_PROCESS_ID'
  }
```

**Comment by I-Want-ToBelieve** [comment]
const cmd = `yabai -m signal --add event=${event} label=${label} action='${yakite} ${action}'`
    $`sh -c ${cmd}`

  work well

**Comment by doubleu88** [comment]
I am having issues with this problem. 

```javascript
const PR_TITLE = "Sample PR title"
const body = "Sample body"

try {
  await $`gh pr create --title=${PR_TITLE} --body=${body}`
} catch (error) {
  throw chalk.red(error.stderr);
}
```

will attempt to run this command
```
gh pr create --title=$'Sample PR title' --body=$'Sample body'
```
which is wrong. 

Expected output:
```
gh pr create --title='Sample PR title' --body='Sample body'
```
Does anyone know of a fix for this?

**Comment by markrreynolds** [comment]
Test function
```js
import { $ } from 'zx';

async function test(name) {
  await $`echo hello: ${name}`;
}

test('world'); // Produces `set -euo pipefail;echo hello: world`
test('not spaces'); // Produces `set -euo pipefail;echo hello: $'not spaces'`
```

**Reply by markrreynolds** [reply]
This appears to be caused by buildCmd inside the vendor "Zurk".

**Reply by markrreynolds** [reply]
Specifically the Zurk borrowed Util method for `quote` doesn't handle spaces and causes the command to be misunderstood.

**Reply by markrreynolds** [reply]
Work around is to create your variable as an array without spaces and pass it into the literal.

```js
import { $ } from 'zx';

async function test(name) {
  await $`echo hello: ${name}`;
}

test(['world']);
test(['not', 'spaces']); // works
```

**Reply by antongolub** [reply]
Hmm... zurk just copy-pasted zx's helper implementation. anyway, could you suggest a patch for the issue?

**Reply by markrreynolds** [reply]
@antongolub Do you happen to know what the intention is behind the `quote` function? There are only two test cases and no docs, so any patch might move the function away from intention.

Update: I stand corrected, there is doc in the primary html docs
https://google.github.io/zx/quotes

**Reply by markrreynolds** [reply]
Well, I am proper confused now. The above, which was failing, is now passing, without any changes. 🤷🏻

**Reply by markrreynolds** [reply]
Found the issue again.
```js
import { $, echo } from 'zx';

async function test(name) {
  const cmd = `echo hello: ${name}`;
  const { stdout } = await $`${cmd}`;
  echo(stdout);
}

test('john smith'); // Produces `set -euo pipefail;echo hello: $'john smith'`
```

**Reply by markrreynolds** [reply]
@antongolub OK, put in a PR. There are a few tests on main that don't seem to work on my machine, but the ones that flex this code are running, and I added a few additional.

https://github.com/google/zx/pull/778

**Reply by antonmedv** [reply]
Please read: https://google.github.io/zx/quotes
Everything works as expected.

**Comment by antonmedv** [comment, marked as answer]
This is a correct behavior. More info at https://google.github.io/zx/quotes

**Reply by markrreynolds** [reply]
I read that document about 4 hours ago (as noted here https://github.com/google/zx/discussions/700#discussioncomment-9104703).

I appreciate that I must have misunderstood the documentation, and that a conflict between the PR and the docs was made.

Before I spend more time on this use case, do you view it as valid?

```js
import assert from 'node:assert'
import { $ }  from './build/core.js'

async function test() {
  const cmd = `echo /tmp/foo bar`
  assert.equal((await $`${cmd}`).stdout.trim(), path)
}

test()
```

**Reply by antonmedv** [reply]
NO. It's not valid. You should not specify cmd as a string. This is considered a bad practice which leads to security issue.

**Reply by markrreynolds** [reply]
Thank. Good luck.
